pkg/autoscaler/autoscaler/util.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package autoscaler
+
+import (
+	"strconv"
+	"time"
+
+	"github.com/pingcap/tidb-operator/pkg/apis/pingcap/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/label"
+	operatorUtils "github.com/pingcap/tidb-operator/pkg/util"
+	appsv1 "k8s.io/api/apps/v1"
+	autoscalingv2beta2 "k8s.io/api/autoscaling/v2beta2"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/utils/pointer"
+)
+
+// checkStsAutoScalingPrerequisites would check the sts status to ensure wouldn't happen during
+// upgrading, scaling
+func checkStsAutoScalingPrerequisites(set *appsv1.StatefulSet) bool {
+	if operatorUtils.IsStatefulSetUpgrading(set) {
+		return false
+	}
+	if operatorUtils.IsStatefulSetScaling(set) {
+		return false
+	}
+	return true
+}
+
+// checkStsAutoScalingInterval would check whether there is enough interval duration between every two auto-scaling
+func checkStsAutoScalingInterval(tc *v1alpha1.TidbCluster, intervalSeconds int32, memberType v1alpha1.MemberType) (bool, error) {
+	if tc.Annotations == nil {
+		tc.Annotations = map[string]string{}
+	}
+	lastAutoScalingTimestamp, existed := tc.Annotations[label.AnnTiDBLastAutoScalingTimestamp]
+	if memberType == v1alpha1.TiKVMemberType {
+		lastAutoScalingTimestamp, existed = tc.Annotations[label.AnnTiKVLastAutoScalingTimestamp]
+	}
+	if !existed {
+		return true, nil
+	}
+	t, err := strconv.ParseInt(lastAutoScalingTimestamp, 10, 64)
+	if err != nil {
+		return false, err
+	}
+	if intervalSeconds > int32(time.Now().Sub(time.Unix(t, 0)).Seconds()) {
+		return false, nil
+	}
+	return true, nil
+}
+
+// checkAutoScalingPrerequisites would check the tidbcluster status to ensure the autoscaling would'n happen during
+// upgrading, scaling and syncing
+func checkAutoScalingPrerequisites(tc *v1alpha1.TidbCluster, sts *appsv1.StatefulSet, memberType v1alpha1.MemberType) bool {
+	if !checkStsAutoScalingPrerequisites(sts) {
+		return false
+	}
+	if memberType == v1alpha1.TiDBMemberType {
+		if tc.Status.TiDB.Phase != v1alpha1.NormalPhase {
+			return false
+		}
+	} else if memberType == v1alpha1.TiKVMemberType {
+		if !tc.Status.TiKV.Synced {
```

Comment:
Why should autoscaler care about the status of tidb and tikv?

Reply:
Because during upgrading or manually scaling, the cpu or other resources would cost much. That's why we need to ensure the auto-scaling wouldn't happen during upgrading or syncing.